File: 프로그래머스/lv3/42628. 이중우선순위큐/이중우선순위큐.py
import heapq

def solution(operations):
    
    # heap 두개활용
    
    minHeap = []
    maxHeap = []
    
    for op in operations:
        m, d = op.split(" ")
        d = int(d)
        if m == "I":
            heapq.heappush(minHeap, d)
            heapq.heappush(maxHeap, -d)
        else:
            if d == -1 and minHeap: # 최소값 삭제
                m = heapq.heappop(minHeap)
                maxHeap.remove(-m)
            elif d == 1 and maxHeap : # 최대값 삭제
                m = heapq.heappop(maxHeap)
                minHeap.remove(-m)
    
    if not maxHeap:  # 최대 힙이 비어있다면, 최소 힙도 비어있어야 함
        minHeap = []
    
    return [-maxHeap[0], minHeap[0]] if maxHeap else [0, 0]

File: 프로그래머스/3/42628. 이중우선순위큐/이중우선순위큐.java
// 1 이상 1,000,000 이하인 문자열 배열
// 우선순위큐 두개
// count로 전체 개수 관리

import java.util.*;

class Solution {
    public int[] solution(String[] operations) {
        int count = 0;
        PriorityQueue<Integer> minPq = new PriorityQueue<>();
        PriorityQueue<Integer> maxPq = new PriorityQueue<>((o1, o2) -> o2-o1);
        
        for (String operation : operations) {
            String[] command = operation.split(" ");
            
            if (command[0].equals("I")) {
                minPq.add(Integer.parseInt(command[1]));
                maxPq.add(Integer.parseInt(command[1]));
                count++;
            }
            else {
                if (command[1].equals("1") && count > 0) {
                    int temp = maxPq.poll();
                    minPq.remove(temp);
                    count--;
                }
                else if (command[1].equals("-1") && count > 0) {
                    int temp = minPq.poll();
                    maxPq.remove(temp);
                    count--;
                }
            }
        }
        
        if (count > 0) {
            return new int[] {maxPq.poll(), minPq.poll()};
        } else {
            return new int[] {0, 0};
        }
    }
}